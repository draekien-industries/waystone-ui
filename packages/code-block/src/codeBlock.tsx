'use client';

import { alpha } from '@theme-ui/color';
import { Card } from '@waystone/card';
import { cssSelectors, outlineCss } from '@waystone/css-presets';
import { Icon } from '@waystone/icon';
import {
  SyntaxHighlighter,
  type SyntaxHighlightStyle as StyleName,
  type SupportedLanguage,
} from '@waystone/syntax-highlighter';
import type { TestIdAttributes } from '@waystone/types';
import { useState, type ReactNode } from 'react';
import { Box, type ThemeUICSSObject } from 'theme-ui';

export type CodeBlockProps = {
  code: string;
  footer?: ReactNode;
  wrapLines?: boolean;
  language?: SupportedLanguage;
  styleName?: StyleName;
} & TestIdAttributes;

const baseDecoratorCss: ThemeUICSSObject = {
  variant: 'text.small',
  letterSpacing: 1.25,
  fontWeight: 'medium',
  position: 'absolute',
  top: 0,
  backgroundColor: 'a-600',
  paddingY: 'extra-small',
  paddingX: 'small',
  textTransform: 'uppercase',
  color: 'b-50',
  boxShadow: 'small',
  borderBottomLeftRadius: 'small',
  borderBottomRightRadius: 'small',
};

export const CodeBlock = ({
  code,
  footer,
  'data-testid': testId,
  ...rest
}: CodeBlockProps) => {
  const [iconName, setIconName] = useState('content_copy');

  const handleCopy = () => {
    navigator.clipboard.writeText(code);
    setIconName('done');
  };

  const handleBlur = () => {
    setIconName('content_copy');
  };

  return (
    <Card data-testid={testId}>
      <Box
        sx={{
          variant: 'text.monospace',
          position: 'relative',
        }}
      >
        {rest.language && (
          <Box
            sx={{
              top: 0,
              left: '1.5rem',
              pointerEvents: 'none',
              ...baseDecoratorCss,
            }}
          >
            {rest.language}
          </Box>
        )}
        <button
          sx={{
            ...baseDecoratorCss,
            right: '1.5rem',
            color: 'text',
            background: 'ghost',
            border: 'none',
            cursor: 'pointer',
            transition: 'all 200ms ease',
            ...outlineCss.base,
            [cssSelectors.hover]: {
              background: 'active',
              ...outlineCss.active,
            },
            [cssSelectors.active]: {
              background: 'active',
              ...outlineCss.active,
            },
            [cssSelectors.focusVisible]: {
              background: 'active',
              ...outlineCss.focused,
            },
          }}
          type="button"
          onClick={handleCopy}
          onBlur={handleBlur}
          aria-label="copy content"
        >
          <Icon>{iconName}</Icon>
        </button>
        <SyntaxHighlighter {...rest}>{code}</SyntaxHighlighter>
        {footer && (
          <Box
            sx={{
              position: 'absolute',
              variant: 'text.caption',
              textAlign: 'center',
              bottom: 0,
              left: 0,
              right: 0,
              backgroundColor: alpha('muted', 0.5),
            }}
          >
            {footer}
          </Box>
        )}
      </Box>
    </Card>
  );
};
