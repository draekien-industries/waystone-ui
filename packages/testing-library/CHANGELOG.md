# @waystone/testing-library

## 9.2.0

### Minor Changes

- 21db598: chore: bump deps and add documentation

### Patch Changes

- Updated dependencies [21db598]
  - @waystone/theme@9.3.0

## 9.1.6

### Patch Changes

- 41664e2: fix: use type imports to improve bundling
- Updated dependencies [41664e2]
  - @waystone/theme@9.2.5

## 9.1.5

### Patch Changes

- 856ba2b: fix: missing dependencies when installing package
- Updated dependencies [856ba2b]
  - @waystone/theme@9.2.4

## 9.1.4

### Patch Changes

- 5b726aa: fix build and dependencies
- Updated dependencies [5b726aa]
  - @waystone/theme@9.2.3

## 9.1.3

### Patch Changes

- 7086019: Update dependencies
- Updated dependencies [7086019]
  - @waystone/theme@9.2.2

## 9.1.2

### Patch Changes

- @waystone/theme@9.2.1

## 9.1.1

### Patch Changes

- Updated dependencies [8f75dd6]
  - @waystone/theme@9.2.0

## 9.1.0

### Minor Changes

- 5c43357: Switch to pnpm and implement minimal syntax highlighter

### Patch Changes

- Updated dependencies [5c43357]
  - @waystone/theme@9.1.0

## 9.0.0

### Major Changes

- 4a19c42: Update package structure

  Created the `@waystone/ui` package for easier installation and created the `@waystone/testing-library`
  package for a preconfigured testing-library render function.

  Uninstall your `@waystone/*` packages, and run the command `yarn add @waystone/ui`
  to update to the latest version. You will need to update the imports to point
  to the new source.

### Patch Changes

- Updated dependencies [4a19c42]
  - @waystone/theme@9.0.0
