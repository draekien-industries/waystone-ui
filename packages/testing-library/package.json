{
  "name": "@waystone/testing-library",
  "version": "10.0.2",
  "private": true,
  "main": "./src/index.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/draekien-industries/waystone-ui"
  },
  "description": "The react component library used by draekien-industries",
  "author": "draekien-industries",
  "sideEffects": false,
  "license": "MIT",
  "files": [
    "src/**"
  ],
  "scripts": {
    "lint": "eslint --ext .tsx,.ts,.jsx,.js ./src/ --fix",
    "lint:release": "eslint --ext .tsx,.ts,.jsx,.js ./src/"
  },
  "devDependencies": {
    "@waystone/tsconfig": "workspace:*",
    "eslint-config-waystone": "workspace:*",
    "eslint-plugin-waystone": "workspace:*",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "dependencies": {
    "@testing-library/jest-dom": "^6.6.3",
    "@testing-library/react": "^16.2.0",
    "@testing-library/user-event": "^14.6.0",
    "@theme-ui/color": "^0.17.1",
    "@theme-ui/css": "^0.17.1",
    "@waystone/theme-provider": "workspace:*",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^29.7.0",
    "theme-ui": "^0.17.1"
  },
  "publishConfig": {
    "access": "public"
  }
}
