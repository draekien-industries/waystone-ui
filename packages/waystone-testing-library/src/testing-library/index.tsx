'use client';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render } from '@testing-library/react';
import { Waystone } from '@waystone/theme-provider';
import type { PropsWithChildren } from 'react';
import type React from 'react';
import { useEffect } from 'react';
import { useColorMode } from 'theme-ui';

const TestProvider = ({ children }: PropsWithChildren) => (
  <Waystone>{children}</Waystone>
);

const DarkModeWrapper = ({ children }: PropsWithChildren) => {
  const [, setColorMode] = useColorMode();

  useEffect(() => {
    setColorMode('dark');
  }, [setColorMode]);

  return <div>{children}</div>;
};

const themedRender: (
  ui: React.ReactElement,
  options?: RenderOptions
) => RenderResult = (ui, options?) =>
  render(ui, { wrapper: options?.wrapper || TestProvider, ...options });

export * from '@testing-library/react';
export * from '@testing-library/user-event';

export { DarkModeWrapper, themedRender as render };
