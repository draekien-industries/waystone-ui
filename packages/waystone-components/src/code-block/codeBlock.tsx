'use client';

import {
  StyleName,
  SupportedLanguage,
  SyntaxHighlighter,
} from '@waystone/syntax-highlighter';
import { useState } from 'react';
import { Icon } from '../icon/icon';
import { copy } from './codeBlock.fx';
import {
  codeBlockCaptionCss,
  codeBlockContainerCss,
  codeBlockCopyButtonCss,
  codeBlockLanguageCss,
} from './codeBlock.styles';

export type CodeBlockProps = {
  /**
   * A flag indicating whether the code block should wrap long lines.
   * @default false
   */
  wordWrap?: boolean;
  /**
   * The text to render as code.
   */
  children: string;
  /** An optional caption to describe the code block. */
  caption?: string;

  /** The code language to use for syntax highlighting. */
  language?: SupportedLanguage;

  /** The name of the syntax highlighting style to use */
  styleName?: StyleName;
};

/**
 * Renders a code block that supports syntax highlighting using HLJS.
 * Refer to https://github.com/react-syntax-highlighter/react-syntax-highlighter/blob/HEAD/AVAILABLE_LANGUAGES_HLJS.MD
 * for the list of available languages.
 * @param props.language - the language of the code inside the codeblock. Used for syntax highlighting.
 * @param props.wordWrap - a flag indicating whether the code block should wrap long lines.
 * @param props.caption - an optional caption to describe the code block.
 * @param props.children - the code to render.
 * @returns a code block component.
 */
export const CodeBlock = ({
  language,
  wordWrap = false,
  caption,
  children,
  styleName = 'tomorrowNightBlue',
}: CodeBlockProps) => {
  const [iconName, setIconName] = useState('content_copy');

  const handleCopy = () => {
    copy(children);
    setIconName('done');
  };

  const handleBlur = () => {
    setIconName('content_copy');
  };

  return (
    <div sx={codeBlockContainerCss}>
      {language && <div sx={codeBlockLanguageCss}>{language}</div>}
      <button
        sx={codeBlockCopyButtonCss}
        type="button"
        onClick={handleCopy}
        onBlur={handleBlur}
      >
        <Icon name={iconName} />
      </button>
      <SyntaxHighlighter
        wrapLines={wordWrap}
        language={language}
        styleName={styleName}
      >
        {children}
      </SyntaxHighlighter>
      {caption && <div sx={codeBlockCaptionCss}>{caption}</div>}
    </div>
  );
};
