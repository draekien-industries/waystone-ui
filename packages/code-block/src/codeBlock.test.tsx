import { render, screen } from '@waystone/testing-library';
import { CodeBlock } from './codeBlock';

describe('<CodeBlock />', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <CodeBlock code="console.log('Hello World');" language="javascript" />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render a footer', () => {
    const { container } = render(
      <CodeBlock code="test" footer="example footer" />
    );

    const footer = screen.getByText('example footer');

    expect(footer).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('accepts a testid', () => {
    render(<CodeBlock code="test" data-testid="code-block" />);

    const codeBlock = screen.getByTestId('code-block');

    expect(codeBlock).toBeInTheDocument();
  });
});
