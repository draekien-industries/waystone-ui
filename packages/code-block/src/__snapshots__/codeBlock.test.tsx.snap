// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CodeBlock /> should match snapshot 1`] = `
<div>
  <div
    class="css-oo4x6t-Card"
  >
    <div
      class="css-ed4rkq-CodeBlock"
    >
      <div
        class="css-l1fdo0-CodeBlock"
      >
        javascript
      </div>
      <button
        aria-label="copy content"
        class="css-11hvwq6-CodeBlock"
        type="button"
      >
        <i
          class="material-icons-outlined css-1ye7pjf-Icon"
        >
          content_copy
        </i>
      </button>
      <pre
        class="css-73wurh-SyntaxHighlighter"
      >
        <code
          class="language-javascript css-dytyo9-SyntaxHighlighter"
        >
          <span
            class="hljs-variable language_ css-g89m8z-Box"
            style="color: rgb(171, 86, 86);"
          >
            <span>
              console
            </span>
          </span>
          <span
            class="css-g89m8z-Box"
          >
            <span>
              .
            </span>
          </span>
          <span
            class="hljs-title function_ css-g89m8z-Box"
            style="color: rgb(136, 0, 0); font-weight: 700;"
          >
            <span>
              log
            </span>
          </span>
          <span
            class="css-g89m8z-Box"
          >
            <span>
              (
            </span>
          </span>
          <span
            class="hljs-string css-g89m8z-Box"
            style="color: rgb(136, 0, 0);"
          >
            <span>
              'Hello World'
            </span>
          </span>
          <span
            class="css-g89m8z-Box"
          >
            <span>
              );
            </span>
          </span>
        </code>
      </pre>
    </div>
  </div>
</div>
`;

exports[`<CodeBlock /> should render a footer 1`] = `
<div>
  <div
    class="css-oo4x6t-Card"
  >
    <div
      class="css-ed4rkq-CodeBlock"
    >
      <button
        aria-label="copy content"
        class="css-11hvwq6-CodeBlock"
        type="button"
      >
        <i
          class="material-icons-outlined css-1ye7pjf-Icon"
        >
          content_copy
        </i>
      </button>
      <pre
        class="css-73wurh-SyntaxHighlighter"
      >
        <code
          class="css-dytyo9-SyntaxHighlighter"
        >
          <span
            class="hljs-built_in css-g89m8z-Box"
            style="color: rgb(57, 115, 0);"
          >
            <span>
              test
            </span>
          </span>
        </code>
      </pre>
      <div
        class="css-klajn0-CodeBlock"
      >
        example footer
      </div>
    </div>
  </div>
</div>
`;
